packages/arb-ts/src/lib/assetBridger/tokenBridger.ts

```diff
@@ -0,0 +1,719 @@
+  gatewayAddr: string
+}
+
+/**
+ * Admin functionality for the token bridge
+ */
+export class AdminTokenBridger extends TokenBridger {
+  /**
+   * Get all the gateway set events on the L1 gateway router
+   * @param l1Provider
+   * @param customNetworkL1GatewayRouter
+   * @returns
+   */
+  public async getL1GatewaySetEvents(
+    l1Provider: Provider,
+    customNetworkL1GatewayRouter?: string
+  ): Promise<GatewaySetEvent['args'][]> {
+    if (this.l2Network.isCustom && !customNetworkL1GatewayRouter) {
+      throw new Error(
+        'Must supply customNetworkL1GatewayRouter for custom network '
+      )
+    }
+
+    const l1GatewayRouterAddress =
+      customNetworkL1GatewayRouter || this.l2Network.tokenBridge.l1GatewayRouter
+
+    const eventFetcher = new EventFetcher(l1Provider)
+    return await eventFetcher.getEvents<L1GatewayRouter, GatewaySetEvent>(
+      l1GatewayRouterAddress,
+      L1GatewayRouter__factory,
+      t => t.filters.GatewaySet()
+    )
+  }
+
+  /**
+   * Get all the gateway set events on the L2 gateway router
+   * @param l1Provider
+   * @param customNetworkL1GatewayRouter
+   * @returns
+   */
+  public async getL2GatewaySetEvents(
+    l2Provider: Provider,
+    customNetworkL2GatewayRouter?: string
+  ): Promise<GatewaySetEvent['args'][]> {
+    if (this.l2Network.isCustom && !customNetworkL2GatewayRouter)
+      throw new Error(
+        'Must supply customNetworkL2GatewayRouter for custom network '
+      )
+
+    const l2GatewayRouterAddress =
+      customNetworkL2GatewayRouter || this.l2Network.tokenBridge.l2GatewayRouter
+
+    const eventFetcher = new EventFetcher(l2Provider)
+    return await eventFetcher.getEvents<L1GatewayRouter, GatewaySetEvent>(
+      l2GatewayRouterAddress,
+      L1GatewayRouter__factory,
+      t => t.filters.GatewaySet()
+    )
+  }
+
+  /**
+   * Register the provided token addresses against the provided gateways
+   * @param l1Signer
+   * @param l2Provider
+   * @param tokenGateways
+   * @returns
+   */
+  public async setGateways(
+    l1Signer: Signer,
+    l2Provider: Provider,
+    tokenGateways: TokenGateway[]
+  ): Promise<ethers.ContractTransaction> {
+    if (!SignerProviderUtils.signerHasProvider(l1Signer)) {
+      throw new MissingProviderArbTsError('l1Signer')
+    }
+
+    const l2GasPrice = await l2Provider.getGasPrice()
+
+    const estimator = new L1ToL2MessageGasEstimator(l2Provider)
+    const { submissionPrice } = await estimator.getSubmissionPrice(
+      // 20 per address, 100 as buffer/ estimate for any additional calldata
+      300 + 20 * (tokenGateways.length * 2),
+      { percentIncrease: BigNumber.from(0) }
+    )
+
+    const l1GatewayRouter = L1GatewayRouter__factory.connect(
+      this.l2Network.tokenBridge.l1GatewayRouter,
+      l1Signer.provider
+    )
+
+    return await l1GatewayRouter.functions.setGateways(
+      tokenGateways.map(tG => tG.tokenAddr),
+      tokenGateways.map(tG => tG.gatewayAddr),
+      0,
+      l2GasPrice,
```

Comment:
Since we're hardcoding max gas to 0, we may as well also leave out l2GasPrice / hardcode that to 0 as well (I think gas price of 0 is fine here @fredlacs ?)

Basically this method creates a retryable ticket, but doesn't autoredeem it / requires manual L2 redemption; maybe we want the name to reflect that (i.e., "createSetGatewaysRetryable Ticket").

Fun little thing would be to have another method that wraps around it, takes in an L2 signer as well, calls it, waits for the L1ToL2Message, and redeems it on L2 directly.

(Alternatively, we could just actually try to estimate the maxGas required, but that's annoying)

Reply:
Ooo maybe for now the best option is include maxGas a param to the method instead of hardcoding it to 0, and we can consume it as we please (this is basically admin only, so not too concerned for now; I have bigger plans r.e. helpers to register custom tokens etc)

Reply:
added the maxgas param for now